### PR TITLE
mgr/dashboard: add multi-cluster management using context switcher in dashboard

### PR DESCRIPTION
**Pull request by aaSharma14**
Allow the user to add a cluster using a form for multi-cluster management and add a context switcher at the top of the navigation bar to allow the user to switch between the clusters that are connected.

[Screencast from 2024-01-19 13-38-37.webm](https://github.com/ceph/ceph/assets/66050535/b12c0af8-9896-42fc-b488-acce7bbc3151)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ parameters:
  prefix: ceph
  numcpus: 1
  memory: 2048
- image: fedora36
+ image: /var/lib/libvirt/images/ceph-dashboard/fedora36
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml:11]: why this change?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        <cd-alert-panel *ngIf="connectionMessage !== 'Connection Verified Successfully' && connectionMessage !== ''"
+                        type="error"
+                        spacingClass="mb-3"
+                        i18n>{{ connectionMessage }}
+        </cd-alert-panel>
+        <cd-alert-panel *ngIf="connectionMessage === 'Connection Verified Successfully'"
+                        type="success"
+                        spacingClass="mb-3"
+                        i18n>{{ connectionMessage }}
+        </cd-alert-panel>
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:19]: I think can be have this message inline to the verify connection button? 

[Verify Connection]  Connecting [ spinner]

[Verify Connection(]button hidden when verified)   Connection verified [Tick green icon]

[Verify Connection] Connection not verified [Cancel red icon]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  templateUrl: './multi-cluster-form.component.html',
+  styleUrls: ['./multi-cluster-form.component.scss']
+})
+export class MultiClusterFormComponent implements OnInit, OnDestroy {
+  readonly endpoints = /^((https?:\/\/)|(www.))(?:([a-zA-Z]+)|(\d+\.\d+.\d+.\d+)):\d{2,4}$/;
+  readonly ipv4Rgx = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/i;
+  readonly ipv6Rgx = /^(?:[a-f0-9]{1,4}:){7}[a-f0-9]{1,4}$/i;
+  remoteClusterForm: CdFormGroup;
+  showToken = false;
+  connectionVerified = false;
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:24]:
```suggestion
  connectionVerified = None;
```

Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:24]: Lets start with `None`, when not checked.
`False` when connection failed
`True` once passed

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 4]
```diff
@@ -0,0 +1,150 @@
+<cd-modal [modalRef]="activeModal">
+  <ng-container i18n="form title"
+                class="modal-title">Connect Cluster
+  </ng-container>
+  <ng-container class="modal-content">
+    <form name="remoteClusterForm"
+          #frm="ngForm"
+          [formGroup]="remoteClusterForm">
+      <div class="modal-body">
+        <cd-alert-panel *ngIf="connectionMessage !== 'Connection Verified Successfully' && connectionMessage !== ''"
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:10]:
```suggestion
        <cd-alert-panel *ngIf="connectionVerified"
```

Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:10]: Why using connection message to check rather than bool?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.scss [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+.error {
+  color: vv.$health-color-error;
+}
+
+.value {
+  font-size: 6.5em;
+}
+
+.fa-wrench {
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.scss:35]: where are we using the CSS?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts [hunk 6]
```diff
@@ -16,8 +16,11 @@ import { CdHelperClass } from '~/app/shared/classes/cd-helper.class';
 import { NotificationType } from '../enum/notification-type.enum';
 import { CdNotificationConfig } from '../models/cd-notification';
 import { FinishedTask } from '../models/finished-task';
-import { AuthStorageService } from './auth-storage.service';
 import { NotificationService } from './notification.service';
+import { MultiClusterService } from '../api/multi-cluster.service';
+import { SummaryService } from './summary.service';
+import { AuthStorageService } from './auth-storage.service';
+// import { SummaryService } from './summary.service';
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts:23]: Just a lonely forgotten comment

**Review by Pegonzal — COMMENTED** [review]
I tested the PR for a bit, and I found the following:
When I switched to the other cluster, it started failing. Could it be because of my setup? Please let me know!
[error.webm](https://github.com/ceph/ceph/assets/106810838/77a917de-a602-4994-a6e3-89106f9fa810)


Some feedback on the **form**:

- Is it needed to have the verify connection button? Why not have everything done with just one button?
- There is some wierd behaviours around the form when checking/unchecking the Token checkbox
[wierd_form_behaviour.webm](https://github.com/ceph/ceph/assets/106810838/b9b20dd7-d102-4eed-88ab-a09f09a89b21)
- It should not be possible to add the same cluster api URL that you are currently on
- I think some tooltips to especify which username and password is required would be nice, what kind of user access is required?
- I created a user without a password and it passed the connectivity verification, but then I can not connect since password is markes a required.
image
image

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 7]
```diff
@@ -58,19 +62,23 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
...
+                    component='dashboard')
 
+            multicluster_config['config'].append({
+                    'name': fsid,
+                    'cluster_alias': cluster_alias,
+                    'url': url,
+                    'token': token
+                })
             Settings.MULTICLUSTER_CONFIG = multicluster_config
             return
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:83]: Also, with the current implementation, what would happend if I send a token and a username and password?
Will it try to authenticate twice

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml (hunk 1 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml]: Updated the PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts (hunk 3 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts]: Thanks @cloudbehl , Updated the PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html (hunk 4 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html]: Updated the PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.scss (hunk 5 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.scss]: Removed

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 8]
```diff
@@ -58,19 +62,23 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
...
-            })
+            try:
+                self._proxy('PUT', url, 'ui-api/multi-cluster/set_cors_endpoint',
+                            payload={'url': hub_url}, token=token)
+                fsid = self._proxy('GET', url, 'api/health/get_cluster_fsid', token=token)
+            except Exception as e:
+                raise DashboardException(
+                    "Could not reach {}".format(url+'api/auth'),
+                    http_status_code=404,
+                    component='dashboard')
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py]: Thanks @Pegonzal , updated the PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                        type="success"
+                        spacingClass="mb-3"
+                        i18n>{{ connectionMessage }}
+        </cd-alert-panel>
+        <div class="form-group row">
+          <label class="cd-col-form-label required"
+                 for="remoteClusterUrl"
+                 i18n>Cluster API URL
+            <cd-helper>Enter the Dashboard API URL here. If you are unsure, you can retrieve it by running the
+              following CLI command: <b>ceph mgr services</b></cd-helper>
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html]: Thanks. Updated

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  </div>
+
+  <span *ngIf="loading"
+        class="d-flex justify-content-center">
+    <i [ngClass]="[icons.large3x, icons.spinner, icons.spin]"></i>
+  </span>
+  <div *ngIf="dashboardClustersMap?.size > 1">
+    <div *ngIf="!loading">
+    </div>
+  </div>
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:50]: This part will be covered in a different PR

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts (hunk 6 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts]: Updated

**Comment by cloudbehl** [comment]
I think we should construct json of storing data in better way. Local cluster(see the image attached) is added everytime I login to cluster. Also we should have username stored as well so in future other user from same cluster can be added.


IMO the JSON should look like this:
```
{
   "current_url":"https://172.20.0.2:11000/",
   "config":{
     "5013cd03-4008-4c45-ae9f-2f0ee0f91a35":  [
      {
         "name":"5013cd03-4008-4c45-ae9f-2f0ee0f91a35",
         "url":"https://172.20.0.2:11000/",
         "cluster_alias":"local-cluster",
         "user": "admin"
      }
    ],
    "79fd48bc-b902-11ee-b0db-52540041612f": [
    {
         "name":"79fd48bc-b902-11ee-b0db-52540041612f",
         "cluster_alias":"cluster1",
         "user": "ankush",
         "url":"https://192.168.100.100:8443/",
      
"token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
      },
    {
         "name":"79fd48bc-b902-11ee-b0db-52540041612f",
         "cluster_alias":"cluster1-abc",
         "user": "admin",
         "url":"https://192.168.100.100:8443/",
      
"token":"UzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
      },
]
   }
}
```


image

**Comment by nizamial09** [comment]
> I think we should construct json of storing data in better way. Local cluster(see the image attached) is added everytime I login to cluster. Also we should have username stored as well so in future other user from same cluster can be added.
> 
> IMO the JSON should look like this:
> 
> ```
> {
>    "current_url":"https://172.20.0.2:11000/",
>    "config":{
>      "5013cd03-4008-4c45-ae9f-2f0ee0f91a35":  [
>       {
>          "name":"5013cd03-4008-4c45-ae9f-2f0ee0f91a35",
>          "url":"https://172.20.0.2:11000/",
>          "cluster_alias":"local-cluster",
>          "user": "admin"
>       }
>     ],
>     "79fd48bc-b902-11ee-b0db-52540041612f": [
>     {
>          "name":"79fd48bc-b902-11ee-b0db-52540041612f",
>          "cluster_alias":"cluster1",
>          "user": "ankush",
>          "url":"https://192.168.100.100:8443/",
>       
> "token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
>       },
>     {
>          "name":"79fd48bc-b902-11ee-b0db-52540041612f",
>          "cluster_alias":"cluster1-abc",
>          "user": "admin",
>          "url":"https://192.168.100.100:8443/",
>       
> "token":"UzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
>       },
> ]
>    }
> }
> ```
> 
> image

In addition to what @cloudbehl said, we'll also need to store the `hub_url` as well.
```
{
   "current_url":"https://172.20.0.2:11000/",
   "hub_url": "https://<hub-ip>:<port>/",
   "config":{
     "5013cd03-4008-4c45-ae9f-2f0ee0f91a35":  [
      {
         "name":"5013cd03-4008-4c45-ae9f-2f0ee0f91a35",
         "url":"https://172.20.0.2:11000/",
         "cluster_alias":"local-cluster",
         "user": "admin"
      }
    ],
    "79fd48bc-b902-11ee-b0db-52540041612f": [
    {
         "name":"79fd48bc-b902-11ee-b0db-52540041612f",
         "cluster_alias":"cluster1",
         "user": "ankush",
         "url":"https://192.168.100.100:8443/",
      
"token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
      },
    {
         "name":"79fd48bc-b902-11ee-b0db-52540041612f",
         "cluster_alias":"cluster1-abc",
         "user": "admin",
         "url":"https://192.168.100.100:8443/",
      
"token":"UzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDU5NDYwNDYsImlhdCI6MTcwNTkxNzI0NiwiaXNzIjoiY2VwaC1kYXNoYm9hcmQiLCJqdGkiOiI4YThmYmI0Mi0yMWNhLTQzOWEtOGM0ZS0wYTFjYzY0ZDFlMDgiLCJ1c2VybmFtZSI6ImFkbWluIn0.N576pNT79ViKEnHJ87MLWbdf9tvRWlkgNP8BeMgIA5c"
      },
]
   }
}
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+              <button class="btn btn-primary"
+                      type="button"
+                      [disabled]="(remoteClusterForm.getValue('showToken') && remoteClusterForm.getValue('apiToken') === '') || (!remoteClusterForm.getValue('showToken') && (remoteClusterForm.getValue('username') === '' || remoteClusterForm.getValue('password') === ''))"
+                      (click)="verifyConnection()">
+                Verify Connection
+              </button>
+            </div>
+          </div>
+        </div>
+      </div>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:139]: I don't like the idea of a separate verify connection button. The `connect cluster` implies that, the connection will be verified as part of that action. So why do we need to do an extra step and force user to click this button?

Instead, when we click the submit button, it'll already check to see if connection can be established. If no connection is established, just trigger an error in the form and display it (either in the form or in notification). I think that would be more than fine.

Moreover, as part of verification, you are actually sending the login request, which means that the same request that we sent when we do Connect Cluster is being used for Verify Connection which duplicates things for unnecessary things.

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 12]
```diff
@@ -39,12 +40,25 @@ def create(self, username, password):
...
+        mgr_map = mgr.get('mgr_map')
+        url = mgr_map.get('services', {}).get('dashboard', None)
         if max_attempt == 0 or mgr.ACCESS_CTRL_DB.get_attempt(username) < max_attempt:
             if user_data:
                 user_perms = user_data.get('permissions')
                 pwd_expiration_date = user_data.get('pwdExpirationDate', None)
                 pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
+            if isinstance(Settings.MULTICLUSTER_CONFIG, str):
+                item_to_dict = json.loads(Settings.MULTICLUSTER_CONFIG)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/auth.py:52]: json.loads will fail if the `Settings.MULTICLUSTER_CONFIG` is not in an expected json format which can happen if someone manually remove the config through dashboard.
```suggestion
               try:
                item_to_dict = json.loads(Settings.MULTICLUSTER_CONFIG)
            except json.JSONDecodeError:
                item_to_dict = {}
```

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 13]
```diff
@@ -39,12 +40,25 @@ def create(self, username, password):
...
+                copy_config = item_to_dict.copy()
+            else:
+                copy_config = Settings.MULTICLUSTER_CONFIG.copy()
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/auth.py:55]:
```suggestion
                multicluster_config = item_to_dict.copy()
            else:
                multicluster_config = Settings.MULTICLUSTER_CONFIG.copy()
```

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 14]
```diff
@@ -39,12 +40,25 @@ def create(self, username, password):
...
+            try:
+                copy_config['config'].append({'name': mgr.get('config')['fsid'], 'url': url, 'cluster_alias': 'local-cluster'})
+            except KeyError:
+                copy_config = {'current_url': url, 'config': [{'name': mgr.get('config')['fsid'], 'url': url, 'cluster_alias': 'local-cluster'}]}
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/auth.py:59]: if you just add a check like
```
if 'config' not in multicluster_config:
    multicluster_config['config'] = {}
```
then you don't need to duplicate the copy_config assigning, which feels like not a clean code.

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 15]
```diff
@@ -58,22 +66,6 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
         else:
             multicluster_config = Settings.MULTICLUSTER_CONFIG.copy()
 
-        if 'hub_url' not in multicluster_config:
-            multicluster_config['hub_url'] = hub_url
-
-        if 'config' not in multicluster_config:
-            multicluster_config['config'] = []
-
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:66]: why are they removed?

if the config is not there in multicluster_config, this will cause errors below when we try to append the config to dict.

Keep in mind that the user can manually delete the multicluster config from the dashboard settings, and that can break the dashboard.

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 16]
```diff
@@ -87,18 +79,66 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
...
+                    'url': url,
+                    'token': token
+                })
+        Settings.MULTICLUSTER_CONFIG = multicluster_config
+
+    @Endpoint('PUT')
+    @UpdatePermission
+    def set_config(self, config: str):
+        if isinstance(Settings.MULTICLUSTER_CONFIG, str):
+            item_to_dict = json.loads(Settings.MULTICLUSTER_CONFIG)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:100]: same thing i pointed above. if not a proper json, this will fail. you need to properly handle the exceptions

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 17]
```diff
@@ -89,16 +97,63 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
...
+            copy_config = item_to_dict.copy()
+        else:
+            copy_config = Settings.MULTICLUSTER_CONFIG.copy()
+        copy_config.update({'current_url': config})
+        Settings.MULTICLUSTER_CONFIG = copy_config
+        return Settings.MULTICLUSTER_CONFIG
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:106]: yup, we don't need the return..

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 18]
```diff
@@ -87,18 +79,66 @@ def auth(self, url: str, name: str, username=None, password=None, token=None, hu
...
+                payload = {
+                    'username': username,
+                    'password': password
+                }
+                content = self._proxy('POST', url, 'api/auth', payload=payload)
+                if 'token' not in content:
+                    return content['detail']
+            except Exception as e:
+                    return 'Connection failed'
+        return 'Connection successful'
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:136]: this feels like duplicate and unnecessary!

Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:136]: this feels like duplicate and unnecessary!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  onSubmit() {
+    const url = this.remoteClusterForm.getValue('remoteClusterUrl');
+    const clusterAlias = this.remoteClusterForm.getValue('clusterAlias');
+    const username = this.remoteClusterForm.getValue('username');
+    const password = this.remoteClusterForm.getValue('password');
+    const token = this.remoteClusterForm.getValue('apiToken');
+
+    this.subs.add(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:89]: why are we using subs.add? can't we go with a normal subscription here?

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.ts [hunk 20]
```diff
@@ -64,6 +64,7 @@ export class LoginComponent implements OnInit {
   }
 
   login() {
+    localStorage.setItem('cluster_api_url', window.location.origin);
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.ts:67]: yup, there will be. we should definitely look for alternatives for this one. But since this setting is also being stored in the dashboard settings and retrieved everytime in the navigation component and then setting (or copying) it to localStorage just for the api-intercepting purposes, its probably okay.

But my concern for token is still present

[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/login/login.component.ts:67]: @aaSharma14 why are we setting the `cluster_api_url` to this value when we login?

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html [hunk 21]
```diff
@@ -30,6 +30,33 @@
 
       <div class="collapse navbar-collapse"
            [ngClass]="{'show': rightSidebarOpen}">
+        <ng-container *ngIf="clustersMap?.size > 1">
+          <ng-container>
+            <div>
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html:35]: yup

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts [hunk 22]
```diff
@@ -116,4 +143,33 @@ export class NavigationComponent implements OnInit, OnDestroy {
...
+      (resp: any) => {
+        localStorage.setItem('cluster_api_url', url);
+        this.selectedCluster = this.clustersMap.get(url);
+        resp['config'].forEach((config: any) => {
+          if (config['name'] === this.selectedCluster['name']) {
+            localStorage.setItem('token_of_selected_cluster', config['token']);
+          }
+        });
+      },
+      () => {},
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:158]: this is just empty, why? and if not needed, can we remove it?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts [hunk 23]
```diff
@@ -28,15 +30,30 @@ export class CdHttpErrorResponse extends HttpErrorResponse {
...
+  localClusterUrl: string;
+  dashboardClustersMap: Map<string, string> = new Map<string, string>();
   constructor(
     private router: Router,
+    public notificationService: NotificationService,
+    private summaryService: SummaryService,
     private authStorageService: AuthStorageService,
-    public notificationService: NotificationService
-  ) {}
+    private multiClusterService: MultiClusterService // private summaryservice: SummaryService
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts:40]:
```suggestion
    private multiClusterService: MultiClusterService
```

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 24]
```diff
@@ -46,9 +48,15 @@ def _proxy(self, method, base_url, path, params=None, payload=None, verify=False
...
     @EndpointDoc("Authenticate to a remote cluster")
-    def auth(self, url: str, name: str, username=None, password=None, token=None, hub_url=None):
+    def auth(self, url: str, cluster_alias: str, username=None, password=None, token=None, hub_url=None):
+        mgr_map = mgr.get('mgr_map')
+        cross_origin_url = mgr_map.get('services', {}).get('dashboard', None)
+        modified_cross_origin_url = cross_origin_url.rstrip('/')
         multicluster_config = {}
 
+        if not url.endswith('/'):
+            url = url + '/'
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:58]: @nizamial09 , but what if a user does it from api?

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html [hunk 25]
```diff
@@ -30,6 +30,33 @@
...
+              <div ngbDropdown
+                   placement="bottom-left"
+                   class="d-inline-block ms-5">
+                <button ngbDropdownToggle
+                        class="btn btn-outline-light cd-context-bar"
+                        i18n-title
+                        title="Selected Cluster:">
+                  <span class="dropdown-text"> {{ selectedCluster.name }} </span>
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html]: Updated

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts (hunk 19 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:91]: we have two subscriptions in the code here..that's why went with subs.add

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:54]: let's not do that for the user now through API. Just doing multi-cluster auth from the API doesn't makes sense for the user.. If needed, and requirement came those can be exposed later. 

Let's do these change separately. i don't want to block this PR further

**Review by nizamial09 — APPROVED** [review]

**Comment by nizamial09** [comment]
make check is failing here. @aaSharma14 please take care of it.

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by nizamial09** [comment]
make check failure unrelated and will be fixed once the feature branch is rebased. so proceeding to merge this
```
_____________________ ERROR collecting alerts/__init__.py ______________________
alerts/__init__.py:2: in <module>
    from .module import Alerts
alerts/module.py:6: in <module>
    from mgr_module import CLIReadCommand, HandleCommandResult, MgrModule, Option
mgr_module.py:1: in <module>
    import ceph_module  # noqa
E   ModuleNotFoundError: No module named 'ceph_module'
```
